lib/core/src/Cardano/Wallet/Primitive/Slotting.hs

```diff
@@ -0,0 +1,262 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Contains tools for converting between @SlotNo@, @EpochNo@, @SlotInEpoch@,
+-- @UTCTime@.
+
+module Cardano.Wallet.Primitive.Slotting
+    ( -- * New api using ouroboros-concensus
+      epochOf
+    , singleEraInterpreter
+    , runQuery
+
+      -- * Old functions
+    , unsafeEpochNo
+    , epochStartTime
+    , epochPred
+    , epochSucc
+    , SlotParameters (..)
+    , slotParams
+    , flatSlot
+    , fromFlatSlot
+    , slotStartTime
+    , slotCeiling
+    , slotFloor
+    , slotAt
+    , slotDifference
+    , slotPred
+    , slotSucc
+    , slotMinBound
+    , slotRangeFromTimeRange
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( ActiveSlotCoefficient (..)
+    , EpochLength (..)
+    , EpochNo (..)
+    , GenesisParameters (..)
+    , Range (..)
+    , SlotId (..)
+    , SlotInEpoch (..)
+    , SlotLength (..)
+    , StartTime (..)
+    , unsafeEpochNo
+    )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Time.Clock
+    ( NominalDiffTime, UTCTime, addUTCTime, diffUTCTime )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+import Ouroboros.Consensus.HardFork.History.EraParams
+    ( EraParams (..), noLowerBoundSafeZone )
+import Ouroboros.Consensus.HardFork.History.Qry
+    ( Qry, runQuery, slotToEpoch )
+import Ouroboros.Consensus.HardFork.History.Summary
+    ( Summary (..), neverForksSummary )
+
+import qualified Cardano.Slotting.Slot as Cardano
+import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
+
+-- -----------------------------------------------------------------------------
+-- New Api using ouroboros-consensus. With the right interpreter, the
+-- calculations don't break on hard-forks.
+
+epochOf :: Cardano.SlotNo -> Qry EpochNo
+epochOf slot = do
+    (e, _, _) <- slotToEpoch slot
+    return $ EpochNo $ fromIntegral $ Cardano.unEpochNo e
+
+-- | An 'Interpreter' for a single era, where the slotting from
+-- @GenesisParameters@ cannot change.
+--
+-- TODO: The type should be changed to @Interpreter@ when we bump
+-- ouroboros-consensus.
```

Comment:
Should we do this? Another PR?

Reply:
No rush. I don't think we need it until #1870 is starting.